### ISSUE TITLE
Evaluation error: Unable to convert R object to Python type.

### ISSUE THREAD

**Issue by TullioM94** [rstudio/reticulate#627]
Hi ! 

I am trying to create my own regularizer function but I am facing some problems. I have looked at the documentation and I have implemented a simple function as suggested: 

L1 <- function(w){
  
  lambda <- 0.0001
  lamb.adjusted <- np$float64(0.0001)
  const.coeff <- np$float64(np$sqrt(k_int_shape(w)[1]))
  penalty <- lamb.adjusted*const.coeff*k_sum(k_abs(w), axis = 1)
  
 return(penalty)
  
} 

which works fine. However, IF I try to define lambda as additional input of the function I receive an error saying that w is an unused argument; do you know why? 

Morevoer, I am trying to implement the same code but using R6 and the code is: 

L1 <- R6::R6Class("Regularizer", inherit = KerasConstraint, 
                  public = list(initialize = function(self, C = 0.){
                                  self$C = np$float64(C)
                                }, 
                                call = function(self, x){
                                  const.coeff <- np$float64(np$sqrt(k_int_shape(weight_matrix)[1]))
                                  penalty <- C*const.coeff*k_sum(k_abs(weight_matrix), axis = 1)
                                  return(penalty)
                                }, 
                                get_config = function(self){
                                  return('C': float(self$l1))
                                }
                                                                        ))
                  
### Then I think we need to wrap this object

L11 <- function(C){
  
  regularizer_l1(L1)
}

But if I then run the code: 

model <- keras_model_sequential() %>% 
  layer_dense(units = 10, activation = "relu", input_shape = dim(train.data)[2]) %>% 
  layer_dense(10, activation = "relu", kernel_regularizer = L11(0.001)) %>% 
  layer_dense(units = 1)

I receive the error in the object of the issue. Do you know why? 

So to summarise: 

1) Why the simple function approach does not work when I impose lambda as an argument to the function? why do I receive the error message of unused w? 

2) Why do I receive the error above when I try to create the object? 

Thank you for your time and your help 

Best

Tullio

**Comment by kevinushey** [comment]
Sorry, but would you be able to provide a minimally reproducible example? Preferably something that I could copy + paste into an R session to reproduce.

**Comment by TullioM94** [comment]
Hi Kevin! 

Thank you for your answer, yes the example is really simple: 

DEFINE AN R6 OBJECT TO BE USED AS RESULARIZED IN KERAS:

L1 <- R6::R6Class("Regularizer", inherit = KerasConstraint,
public = list(initialize = function(self, C = 0.){
self$C = k_cast_to_floatx(C)
},
call = function(self, x){
const.coeff <- np$float64(np$sqrt(k_int_shape(weight_matrix)[1]))
penalty <- Cconst.coeffk_sum(k_abs(weight_matrix), axis = 1)
return(penalty)
},
get_config = function(self){
return('C': float(self$l1))
}
))

WRAPPING THE FUNCTION: 

L11 <- function(C){
regularizer_l1(L1)
}

FITTING A FNN USING THE KERAS PACKAGE

model <- keras_model_sequential() %>% 
  layer_dense(units = 10, activation = "relu", input_shape = dim(train.data)[2]) %>% 
  layer_dense(units = 11, activation = "relu", kernel_regularizer = L11(0.0001)) %>% 
  layer_dense(units = 11, activation = "relu", kernel_regularizer = L11(0.02)) %>% 
  layer_dense(units = 11, activation = "relu", kernel_regularizer = L1(0.1)) %>% 
  layer_dense(units = 1)

model %>% compile(loss = "mse", optimizer = optimizer_rmsprop(), metrics = list("mean_absolute_error"))

history <- model %>% fit(train.data, train.labels, epochs = 1000, validation_split = 0.2, verbose = 0) 


Thank you again for your time and help 


Tullio

**Comment by skeydan** [comment]
Hi,

please make use of github formatting to display the complete executable and runnable code, like so

```
L11 <- function(C){
regularizer_l1(L1)
}
```

Also, please check the code really is runnable as-is -- for example, when I copy-paste the above, `dim(train.data)` will throw an error :-)

**Comment by TullioM94** [comment]
```
library(keras)
library(tensorflow)
library(tfestimators)
library(tfdatasets)
library(reticulate)
library(dplyr)

sys <- import("sys")
np <- import("numpy")

boston_housing <- dataset_boston_housing()
c(train.data, train.labels) %<-% boston_housing$train
```
```
L1 <- R6::R6Class("Regularizer", inherit = KerasConstraint,
public = list(initialize = function(self, C = 0.){
self$C = k_cast_to_floatx(C)
},
call = function(self, x){
const.coeff <- np$float64(np$sqrt(k_int_shape(weight_matrix)[1]))
penalty <- Cconst.coeffk_sum(k_abs(weight_matrix), axis = 1)
return(penalty)
},
get_config = function(self){
return('C': float(self$l1))
}
))
```

```
L11 <- function(C){
regularizer_l1(L1)
}
```
```
model <- keras_model_sequential() %>%
layer_dense(units = 10, activation = "relu", input_shape = dim(train.data)[2]) %>%
layer_dense(units = 11, activation = "relu", kernel_regularizer = L11(0.0001)) %>%
layer_dense(units = 11, activation = "relu", kernel_regularizer = L11(0.02)) %>%
layer_dense(units = 11, activation = "relu", kernel_regularizer = L1(0.1)) %>%
layer_dense(units = 1)

model %>% compile(loss = "mse", optimizer = optimizer_rmsprop(), metrics = list("mean_absolute_error"))

history <- model %>% fit(train.data, train.labels, epochs = 1000, validation_split = 0.2, verbose = 0)
```

**Comment by skeydan** [comment]
The R6 object doesn't look quite right to me...

For example, this line

```
penalty <- Cconst.coeffk_sum(k_abs(weight_matrix), axis = 1)
```
is broken, and what is `weight_matrix`? Where is `self$l1` defined? Also, this 
```
return('C': float(self$l1))
```

isn't valid R code...

I'd  suggest you start by a simple do-nothing class that throws no error and then gradually add functionality to that.

**Comment by TullioM94** [comment]
Hi! 

Ok, so my main objective is to create a custom regularizer. If I look at the documentation in Keras I have understood that there are two options: either with a function or a callable object. If I use a custom function as below, I have to specify the tuning parameter inside the function as below: 

```
L1 <- function(w){

  lamb <- np$float64(0.001)
  const.coeff2 <- np$float64(np$sqrt(k_int_shape(w)[1]))  
  penalty <- ((lamb*const.coeff)*(k_sum((tf$norm(w, ord = "euclidean", axis = 1L)))) 
  return(penalty)
  
} 

```

The other option that I have is to create a callable object, and as there is no documentation on the regularizer. I am using the Github code for the regularizer of the keras package in python. The python code, does not require as input of the callable object the weights, and I am trying to replicate the code:

```
Regularizer <- R6::R6Class("Regularizer", public = list(
  C = NULL,
  x = NULL,
  initialize = function(C = 0.){
    self$C <- np$float64(C)
  }, 
  call = function(x){
    const_coeff <- np$float64(np$sqrt(k_int_shape(x)[1]))
    final <- self$C*const_coeff*k_sum(k_sqrt(k_sum(k_square(x), axis = 1)))
    return(final)
  } 
  
))
```
If I test the object above using the code below, it works: 

```
w <- tf$random_normal(shape(13, 50), stddev = 0.1)

test1 <- Regularizer$new(C = 3)
test2 <- test1$call(w)
sess = tf$Session()
sess$run(test2)
```

If then I try to adapt the object above for the Keras case: 

```
Regularizer <- R6::R6Class("Regularizer", public = list(
  C = NULL,
  initialize = function(self, C = 0.){
    self$C <- np$float64(C)
  }, 
  call = function(self, x){
    const_coeff <- np$float64(np$sqrt(k_int_shape(x)[1]))
    final <- self$C*const_coeff*k_sum(k_sqrt(k_sum(k_square(x), axis = 1)))
    return(final)
  } , 

 get_config = function(self){
 return({'C' : np$float(self$l1)})

}
  
))

```

And then I add it to the keras model I have the error message: "Could not find function Regularizer"

```
model <- keras_model_sequential() %>% 
  layer_dense(units = 10, activation = "relu", input_shape = dim(train.data)[2]) %>% 
  layer_dense(units = 11, activation = "relu", kernel_regularizer = Regularizer(0.001)) %>% 
  layer_dense(units = 11, activation = "relu") %>% 
  layer_dense(units = 11, activation = "relu") %>% 
  layer_dense(units = 1)

```

So, based on the Github code of Keras regularizer in Rstudio I try to create a function: 

```
regularizer_l11 <- function(l = 0.01){
  Regularizer(C = l)
}
```

And If I use the code: 

```
model <- keras_model_sequential() %>% 
  layer_dense(units = 10, activation = "relu", input_shape = dim(train.data)[2]) %>% 
  layer_dense(units = 11, activation = "relu", kernel_regularizer = regularizer_l11(0.001)) %>% 
  layer_dense(units = 11, activation = "relu") %>% 
  layer_dense(units = 11, activation = "relu") %>% 
  layer_dense(units = 1)

```

It returns the same error. As the error is different, if you think that this issue is no longer valid in the reticulate Github I will move it to Keras. 


Thank you for your time and help 



Tullio

**Comment by TullioM94** [comment]
Sorry for the second post but I coded another option that returns the original error. From above I can define:

```
Regularizer <- R6::R6Class("Regularizer", public = list(
  C = NULL,
  initialize = function(C = 0.){
    self$C <- np$float64(C)
  }, 
  call = function(x){
    const_coeff <- np$float64(np$sqrt(k_int_shape(x)[1]))
    final <- self$C*const_coeff*k_sum(k_sqrt(k_sum(k_square(x), axis = 1)))
    return(final)
  } 
  
    ))


regularizer_l11 <- function(l = 0.01){
  Regularizer$new(C = l)
  
}


model <- keras_model_sequential() %>% 
  layer_dense(units = 10, activation = "relu", input_shape = dim(train.data)[2]) %>% 
  layer_dense(units = 11, activation = "relu", kernel_regularizer = regularizer_l11(0.001)) %>% 
  layer_dense(units = 11, activation = "relu") %>% 
  layer_dense(units = 11, activation = "relu") %>% 
  layer_dense(units = 1)

```

It returns the error: "Evaluation error: Unable to convert R object to Python type." But it does not return the error of "Function not found mentioned above". 



Thank you 


Best, 


Tullio

**Comment by skeydan** [comment]
Hi,
if I understand correctly, the easiest way here is to go with a simple function approach, like so (you create a function to create that function :-)):

```
create_l1 <- function(c) {
  function(w) {
    x <- 1
    x * c
  }
}

model <- keras_model_sequential() %>% 
  layer_dense(units = 10, activation = "relu", input_shape = 2) %>% 
  layer_dense(units = 11, activation = "relu", kernel_regularizer = create_l1(2.5)) %>% 
  layer_dense(units = 1)

model %>% compile(loss = "mse", optimizer = "adam")

model %>% fit (
  x = matrix(1:10, ncol = 2),
  y = matrix(1:5, ncol = 1)
)

```

Does that work for you? (As an aside, you do not need to use `np$` inside R code; R to Python conversion gets taken care of by `reticulate` automatically.)

**Comment by TullioM94** [comment]
Hi! 

yes you are right that was the easiest approach and it works perfectly! Ah ok thanks for the np tip! 

I will leave the R6 object aside then, if I find the solution I will post it here! 

Thank you again 

Best, 


Tullio

**Comment by skeydan** [comment]
you're welcome!

**Comment by zeshan304** [comment]
@TullioM94 I am facing the same problem for converting the r object to python. Did you find any solution of your problem?

**Comment by t-kalinowski** [comment]
@zeshan304 This is an old issue, can you please open a new issue, ideally with a code snippet I can copy+paste to reproduce the error locally.

**Comment by zeshan304** [comment]
> 
Thanks
@t-kalinowski  i have open new issue #1336
